### PR TITLE
Elevate all known options into AmpProject\AmpWP\Option interface

### LINKED ISSUE #4526: Elevate all known options into the `AmpProject\AmpWP\Option` interface
## Feature description

When I introduced the CSS transient caching monitor, I needed to add a new option to the `AMP_Options_Manager`, and so I declared it in a new `AmpProject\AmpWP\Option` interface.

I suggest elevating all known options into this interface and referring to them via the interface's constants.

Benefits:

- Typos are caught immediately, instead of just returning inconsistent results, as `Option::SOME_AMP_OPTION` is a language construct that both the IDE and static analysis will immediately verify.

- It provides convenient auto-completion of option names.

- We have one place where there's a clear overview of what options exist.

- We can use this file for extensive comments, to document current and deprecated options, their expected types and their usage.

- It formalizes the public interface that we provide for 3P code, instead of relying on arbitrary strings.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* All known current and deprecated options should have a constant in the `AmpProject\AmpWP\Option` interface.

## Implementation brief

* Extract option name strings into constants within `AmpProject\AmpWP\Option`
* Replace option name strings throughout the code with the relevant constants

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by schlessera**
## Summary

- Add all known options into `AmpProject\AmpWP\Option` interface.
- Document options within `AmpProject\AmpWP\Option`.
- Adapt option usage to reference the `AmpProject\AmpWP\Option` interface instead of using hard-coded strings.

Image 2020-04-07 at 9 34 21 AM

Options can now be safely auto-completed.
Deprecated options are shown in a strike-through font to make their deprecation obvious.

Image 2020-04-07 at 9 37 32 AM
When requesting help while the cursor is on such a constant, you get additional contextual information about the specific option.

Fixes #4526

## Checklist

- [x] My pull request is addressing an [open issue](https://github.com/ampproject/amp-wp/contributing/project-management.md#life-of-an-issue) (please create one otherwise).
- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/contributing/engineering.md#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/contributing/engineering.md) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by schlessera** [comment]
@westonruter, @kienstra, @pierlon It would be great to add any additional missing documentation to these option constants, so we'll have one single place that we can refer to for historic knowledge.

**Review by kienstra — APPROVED** [review]
**Looks Good**

Hi @schlessera,
This looks good!

The only place I noticed that could probably use a constant is:

https://github.com/ampproject/amp-wp/blob/014bbbba04bea102387f6faaa6b56ab8bf0e1e0a/includes/amp-helper-functions.php#L951

Also, basic smoke testing looks good, like changing the template mode.

File: includes/options/class-amp-options-manager.php [hunk 1]
```diff
@@ -115,28 +117,28 @@ public static function get_options() {
...
 			 * @since 1.5.0
 			 */
-			$options['story_templates_version'],
-			$options['story_export_base_url'],
-			$options['story_settings'],
-			$options['enable_amp_stories'], // This was added in 1.2-beta and later migrated into the `experiences` option.
+			$options[ Option::STORY_TEMPLATES_VERSION ],
+			$options[ Option::STORY_EXPORT_BASE_URL ],
+			$options[ Option::STORY_SETTINGS ],
+			$options[ Option::ENABLE_AMP_STORIES ], // This was added in 1.2-beta and later migrated into the `experiences` option.
```
Comment (pierlon) [inline includes/options/class-amp-options-manager.php]: I suppose this comment can be moved or added to to the option's docblock.

**Review by pierlon — CHANGES_REQUESTED** [review]
Great work here @schlessera! As for the failing tests, the `\AmpProject\AmpWP\Option` interface needs to be imported in:

- `\Test_AMP_CLI_Validation_Command::class`
- `\Test_AMP_Post_Meta_Box::class`
- `\Test_AMP_Validation_Manager::class`

Also, I suppose the todo in the `Option` interface can be removed:

```diff
--- src/Option.php	(revision a1ddd47164d0dbd5fe80457da98aee69c4d3082a)
+++ src/Option.php	(date 1586343490276)
@@ -10,8 +10,6 @@
 /**
  * An interface to share knowledge about options stored in the AMP Options Manager.
  *
- * @todo Other options used throughout the plugin should use constants in this interface as well.
- *
  * @package AmpProject\AmpWP
  */
 interface Option {
```

File: includes/options/class-amp-options-manager.php (hunk 1 above)
Comment (schlessera) [reply includes/options/class-amp-options-manager.php]: Fixed via https://github.com/ampproject/amp-wp/pull/4545/commits/623bcbbeb1ff564269fe73f21459ea3a19089954

**Review by pierlon — APPROVED** [review]

**Review by westonruter — APPROVED** [review]

**Pull request merged**